Entropy Stack update

rivos-eblot opened the pull request:
This update contains mostly EDN IP update which has been heavily reworked based on HW simulation (Verilator traces) rather than on EDN documentation, as it has been done for the initial implementation.

It also contains several updates and fixes for the CSRNG and EntropySrc, but those IPs still require some updates, which will be delivered in a forth coming PR.

rivos-eblot committed:
[ot] hw/opentitan: ot_edn: rework EDN to match new HW implementation

- new internal FSM states
- new registers (HW vs. SW commands)
- new alert on CSRNG error
- reworked instantiation/uninstantiation sequences
- SW command check whether a GENERATE command is transmitted, and preset the expected
  entropy packet count as done with Boot and Auto GENERATE commands.
- CSRNG error management
- bug fixing

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_csrng: rework CSRNG to match new implementation

rework CSRNG public API:
- use enumerated CSRNG results, which matches the actual HW
- only accepts 32-bit word as input command, not a buffer. This better
  replicates the actual EDN-CSRNG HW interface
- accepts up to 16 command words, such as the HW, even if only up to 13
  are meaningful

features update:
- detect invalid FLAG0 multibit bool in ACMD commands
- fix reseed_counter management based on HW wave observation
- raise IRQ + ALERT for the same "fatal" cause
- make ALERT edge-triggered rather than level-triggered
- duplicate internal state value into RESEED_COUNTER registers
- implement RESEED_INTERVAL check
- implement FIPS_FORCE, INT_STATE_READ_ENABLE, NT_STATE_READ_ENABLE_REGEN
  register features
- update REGWEN protected registers
- do not try any longer to enforce the documented entropy stack power cycle:
  only emit a warning about discrepancy between the recommended, documented
  sequence and the actual guest behavior, but do not consider the invalid
  sequence as an error anymore
- flush the application CMD FIFO queue on client disconnect

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_fifo32: update function signatures with const

This departs from QEMU's fifo8 implementation, but enables to use the
OT FIFO from a const context.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: fix timing hint computation.

A timer may be pending but already expired, in which case the remaining
time is no longer valid.

Do not try any longer to enforce the documented entropy stack power cycle:
only emit a warning about discrepancy between the recommended, documented
sequence and the actual guest behavior, but do not consider the invalid
sequence as an error anymore.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv: ot_earlgrey: remove invalid EDN reference from LC_CTRL

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] .gitlab-ci.d: opentitan: update BM tests

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: update register definitions

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv: ot_earlgrey: warn only once on dummy devices access

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_common: make ot_common_configure_device_opts public

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv: ot_earlgrey: add a Verilator-compliant clock mode

This enables runnning applications built for Verilator simulator on QEMU.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_aes: add trace for selected reseed rates

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

loiclefort commented on hw/opentitan/ot_csrng.c:
```diff
@@ -666,43 +763,51 @@ static void ot_csrng_drng_update(OtCSRNGInstance *inst)
     drng->seeded = false;
 
     if (!flag0) {
         if (!inst->parent->es_available) {
```
could use `s->es_available` here

rivos-eblot replied:
true, done

loiclefort commented on hw/opentitan/ot_csrng.c:
```diff
@@ -955,22 +1052,28 @@ static void ot_csrng_handle_enable(OtCSRNGState *s)
+                    sts = CSRNG_STATUS_INVALID_CMD_SEQ;
+                }
+                if (sts) {
+                    // temporary till implemented
+                    qemu_log("%s: UNIMP: CSRNG ERROR STATUS\n", __func__);
```
maybe `qemu_log_mask(LOG_UNIMP, ...`?

rivos-eblot replied:
actually remove, it was part of left over traces while comparing with the HW waves

loiclefort commented on hw/opentitan/ot_csrng.c:
```diff
@@ -1032,17 +1137,19 @@ static void ot_csrng_complete_command(OtCSRNGInstance *inst, int res)
     } else {
         CHANGE_STATE(s, CSRNG_ERROR);
+        // temporary till implemented
+        qemu_log("%s: UNIMP: CSRNG ERROR STATUS\n", __func__);
```
same here, `qemu_log_mask(LOG_UNIMP, ...`

rivos-eblot replied:
same as well

loiclefort commented on hw/opentitan/ot_csrng.c:
```diff
@@ -1345,20 +1477,21 @@ static int ot_csrng_handle_command(OtCSRNGState *s, unsigned slot)
     if (!s->enabled) {
         qemu_log_mask(LOG_GUEST_ERROR, "%s: not enabled\n", __func__);
-        return -1;
+        // @todo check error code
```
left over `@todo` here

rivos-eblot replied:
removed

loiclefort commented on include/hw/opentitan/ot_csrng.h:
```diff
@@ -80,15 +90,16 @@ typedef void (*ot_csrng_genbit_filler_fn)(void *opaque, const uint32_t *bits,
  * @req_sts the IRQ to signal once a command is completed. The IRQ level signal
  *          the completion status of the command: 0 indicates a sucessful
- *          completion, non-zero a failed one.
+ *          completion, non-zero a failed one. Ignored if filler is NULL.
```
`s/filler/fn/` or `s/filler/filler function/` maybe? or rename the function parameter to `filler`?

rivos-eblot replied:
renamed `fn` as `filler_fn`

rivos-eblot committed:
[ot] hw/opentitan: ot_aes: delay status_output_valid bit

When AES forces a PRNG reseed, the EDN may take some time to deliver an entropy packet.
However, unit tests assume that as soon as data out is available, once output data have been read,
the AES IP should return to the IDLE state. IDLE state cannot be reached till a PRNG reseed is
ongoing. This means that the output data should not be made available to the vCPU until any PRNG
reseed has completed.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot commented:
Updated EDN: GENERATE command header was not parsed in `SW_PORT_MODE` (as it's done for Boot and Auto mode), resulting of 0 expected entropy packet, deadlocking the CSRNG comm channel. Fixed.

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: improve trace messages

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_csrng: rework FIPS compliance tracking

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_otbn: enable reporting errors without tracing execution.

Tracing execution massively slows down OTBN execution. Add a new option to
trace execution, while leaving logfile option to only report execution errors
when trace logging is not enabled.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

luismarques commented:
It's going to take me a while to mentally digest all of the changes in this PR, so a review from me might take a bit, but I don't want to delay this too much. Maybe we could occasionally do review meetings where we go over the code together?

rivos-eblot committed:
[ot] hw/opentitan: ot_sensor: renamed as ot_sensor_eg

OT Sensor IP is top-specific

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv: ot_darjeeling: remove OT_SENSOR

OT_SENSOR IP is no longer present on Darjeeling

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_sensor_eg: update register definitions

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] python/qemu: ot.pyot.executer: fix a bug when a single ROM is used.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_edn: rework FIPS compliance tracking

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_otbn: reformat code

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_edn: add a sanity check on command sequence.

A new command should never be emitted if there is a previous one on going.
Remove useless command buffer wipes.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_edn: fix issues with disablement handling.

EDN_ENABLE is not a boolean, but a multibit boolean.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

AlexJones0 commented on hw/riscv/ot_earlgrey.c:
```diff
@@ -183,10 +184,16 @@ enum OtEGBoardDevice {
 /* EarlGrey/CW310 AON clock is 250 kHz */
 #define OT_EG_AON_CLK_HZ 250000u
 
+/* Verilator Core clock is 500 kHz */
+#define OT_VERILATOR_CORE_CLK_HZ 500000u
+/* Verilator Peripheral clock is 125 kHz */
+#define OT_VERILATOR_PERIPHERAL_CLK_HZ ((OT_VERILATOR_CORE_CLK_HZ) / 4u)
+/* Verilator AON clock is 125 kHz */
+#define OT_VERILATOR_AON_CLK_HZ OT_VERILATOR_PERIPHERAL_CLK_HZ
```
Nit: `s/OT_VERILATOR/OT_EG_VERILATOR/` here for naming consistency?

rivos-eblot replied:
Ok

rivos-eblot committed:
[ot] hw/opentitan: ot_edn: replace legacy reset with Resettable implementation

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_ibex_wrapper: rename OtIbexWrapperStateClass

This naming better follows other OT class naming conventions.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_common: redefine some OBJECT macro with an explicit class name

QEMU assumes that the class name should be built from the instance name, however this is
not even used constantly across QEMU files. Only when abstract classes are used this
auto naming scheme is applied.

Redefine two similar macros (from the original ones) to provide the actual object class
name.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_ibex_wrapper: replace legacy reset with Resettable implementation

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_ibex_wrapper: add reset_exit implementation

Upon reset the data will be invalid with a new EDN request pending.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot commented:
> Maybe we could occasionally do review meetings where we go over the code together?
Sure. I could take a couple of hours for such meetings, but I'm afraid it could not be much more, given how many IPs needs to be rewritten.

AlexJones0 commented on hw/riscv/ot_earlgrey.c:
```diff
@@ -1392,16 +1405,78 @@ static void ot_eg_soc_reset_exit(Object *obj, ResetType type)
+        case OT_EG_PERIPHERAL_CLK_HZ:
+            pclk = OT_VERILATOR_PERIPHERAL_CLK_HZ;
+            break;
+        case OT_EG_AON_CLK_HZ:
+            continue;
```
I think I'm missing some context here - is there a reason we don't update the AON clock for Verilator as well?

rivos-eblot replied:
Looks like I forgot... Thanks.

rivos-eblot committed:
[ot] hw/opentitan: ot_csrng: fix connection management

1. inverted detection of a previous EDN connection
2. the filler_fn function now tracks whether an EDN is connected or not,
   the instantiation status was a left over from the initial implementation

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_edn: delayed mode change is legit, not an error

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_csrng: add missing write access to 2 new registers

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: add missing read access to 1 register

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot commented:
Some progress since last draft:
```
+----------------------------+---------+-----------+--------+----------------------------------------+
| Name                       |  Result |      Time | Icount |                                  Error |
+============================+=========+===========+========+========================================+
| csrng_edn_concurrency_test |    PASS | 145839 ms |      6 |                                        |
+----------------------------+---------+-----------+--------+----------------------------------------+
| csrng_kat_test             |    PASS |    267 ms |      6 |                                        |
+----------------------------+---------+-----------+--------+----------------------------------------+
| csrng_smoketest            |    PASS |    206 ms |      6 |                                        |
+----------------------------+---------+-----------+--------+----------------------------------------+
| edn_auto_mode              |    PASS |   9900 ms |      6 |                                        |
+----------------------------+---------+-----------+-------------------------------------------------+
| edn_boot_mode              |    FAIL |    200 ms |      6 | CHECK-STATUS-fail: Internal:["OBT";47] |
+----------------------------+---------+-----------+--------+----------------------------------------+
| edn_kat                    | TIMEOUT |  30002 ms |      6 |                                        |
+----------------------------+---------+-----------+--------+----------------------------------------+
| edn_sw_mode                |    PASS |   1505 ms |      6 |                                        |
+----------------------------+---------+-----------+--------+----------------------------------------+
```
`edn_boot_mode` failure is due to the OTBN not receiving the expected FIPS compliance status from the EDN, on the second round (first round is ok)

rivos-eblot commented:
It might be worth splitting this PR once entropy test are passing. Entropy tests depend on so many IPs that is difficult to predict the proper set of IPs to update for now.

AlexJones0 commented on include/hw/opentitan/ot_csrng.h:
```diff
@@ -80,23 +90,25 @@ typedef void (*ot_csrng_genbit_filler_fn)(void *opaque, const uint32_t *bits,
  *         receive entropy, i.e. genbits_ready
  */
 qemu_irq ot_csnrg_connect_hw_app(OtCSRNGState *s, unsigned app_id,
                                  qemu_irq req_sts, ot_csrng_genbit_filler_fn fn,
```
Nit: can you also change `fn` -> `filler_fn` here so the docs match the parameter names? You made this change in the definition in `ot_csrng.c` but seem to have missed it in this declaration.

rivos-eblot replied:
Sure.

AlexJones0 commented on hw/opentitan/ot_csrng.c:
```diff
@@ -334,16 +351,25 @@ struct OtCSRNGState {
+    [CSRNG_GENERATE_REQ]  = 0b00100100, // 0x24: generate request (takes adata? and Key,V,RC)
+    [CSRNG_UPDATE_PREP]   = 0b00110001, // 0x31: update prep
+    [CSRNG_UPDATE_REQ]    = 0b00100100, // 0x24: update request (takes adata and Key,V,RC)
+    [CSRNG_UNINSTANT_PREP]= 0b11110110, // 0xf6: uninstantiate prep
```
Nit: formatting is missing a space here?

rivos-eblot replied:
Ok. Plus I did not notice the comments reaching over 80 colums.

AlexJones0 commented on hw/opentitan/ot_csrng.c:
```diff
@@ -666,59 +762,67 @@ static void ot_csrng_drng_update(OtCSRNGInstance *inst)
                                                       "not ready",
                                             res);
-            return res < 0 ? (res == -2 ? CSRNG_CMD_STALLED : CSRNG_CMD_ERROR) :
+            return res < 0 ? (res == -2 ? CSRNG_CMD_STALLED :
+                                          CSRNG_CMD_RESEED_CNT_EXCEEDED) :
```
I'm not sure if there's just some documentation out of sync, but this doesn't seem to entirely match up with the documentation on `get_random_values` in `ot_random_src.h`? Is something out of date here?

rivos-eblot replied:
You are right, I need to double check

rivos-eblot replied:
removed/reworked in last-but-one commit

AlexJones0 commented on hw/opentitan/ot_csrng.c:
```diff
@@ -666,59 +762,67 @@ static void ot_csrng_drng_update(OtCSRNGInstance *inst)
-
+    drng->reseed_counter = 0u;
     drng->seeded = true;
+    drng->force_fips = (bool)(s->regs[R_FIPS_FORCE] >> slot);
```
Maybe I'm misunderstanding, but shouldn't this be something like
```suggestion
    drng->force_fips = (bool)((s->regs[R_FIPS_FORCE] >> slot) & 0x1);
```
as otherwise for example forcing the FIPS of just instance 2 would also lead this to be true for slot 1, despite the fact that instance 1 itself might be disabled? I'm not too familiar with the CSRNG so I may just not have a full understanding here.

rivos-eblot replied:
Oops. good catch

AlexJones0 commented on hw/opentitan/ot_csrng.c:
```diff
@@ -1507,10 +1622,12 @@ static void ot_csrng_command_scheduler(void *opaque)
+        break;
     default:
-        ot_csrng_complete_command(inst, res);
+        trace_ot_csrng_reject_command(slot, command, res);
+        ot_csrng_complete_command(inst, -res);
```
I'm finding the logic slightly confusing to follow here - would it be possible to add a short comment documenting this case? Why `-res`?

rivos-eblot replied:
Sure. I've added an comment and a description for `OtCSRNDCmdResult`:

```h
/*
 * Command execution result.
 *
 * Zero value signals the success and immediate completion of a command
 * Positive values signal a deferred completion of a command
 * Negative values signal an error, encoding a OtCSRNGCmdStatus value
 */
typedef enum {
    /* entropy stack is stalled */
    CSRNG_CMD_STALLED = -CSRNG_STATUS_COUNT,
    /* command errors, not recoverable */
    CSRNG_CMD_RESEED_CNT_EXCEEDED = -CSRNG_STATUS_RESEED_CNT_EXCEEDED,
    CSRNG_CMD_INVALID_CMD_SEQ = -CSRNG_STATUS_INVALID_CMD_SEQ,
    CSRNG_CMD_INVALID_GEN_CMD = -CSRNG_STATUS_INVALID_GEN_CMD,
    CSRNG_CMD_INVALID_ACMD = -CSRNG_STATUS_INVALID_ACMD,
    /* command completed ok */
    CSRNG_CMD_OK = 0,
    /* command cannot be executed for now */
    CSRNG_CMD_RETRY = 1,
    /* command completion deferred */
    CSRNG_CMD_DEFERRED = 2,
} OtCSRNDCmdResult;
```

rivos-eblot committed:
[ot] python/qemu: ot.eflash.gen: fix a bug when storing an ELF file

When an ELF file is used to store an app in the flash image, and not
binary physical file exists, the binary data is automatically generated
but no match should be attempted between the ELF metadata and the
generated binary content.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: ignore generation identifier

The documented rules

1. ENTROPY_SRC may only be disabled if CSRNG is disabled.
3. Once disabled, CSRNG may only be re-enabled after ENTROPY_SRC has been disabled and re-enabled.

no longer apply: remove the related checks.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_entropy_src: fix initial values

They had not been updated when registers had.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_otbn: fix FIPS compliance management

FIPS status should be reset after each entropy packet is delivered to the OTBN core.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot commented:
```
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| Name                                                        |  Result |      Time | Icount |                                                                       Error |
+=============================================================+=========+===========+========+=============================================================================+
| aes_entropy_test                                            |    PASS |    307 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| aes_force_prng_reseed_test                                  | TIMEOUT |  10001 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| aes_idle_test                                               |    PASS |    315 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| aes_masking_off_test                                        | TIMEOUT |  10001 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| aes_prng_reseed_test                                        | TIMEOUT |  10001 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| aes_smoketest                                               |    PASS |    374 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| aes_stall_test                                              |    FAIL |    339 ms |      6 | CHECK-fail: ERROR: AES module not STALLED as expected after loading block 1 |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| csrng_edn_concurrency_test                                  |    PASS | 171378 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| csrng_kat_test                                              |    PASS |    268 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| csrng_smoketest                                             |    PASS |    207 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| edn_auto_mode                                               |    PASS |  14308 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| edn_boot_mode                                               |    PASS |    846 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| edn_kat                                                     | TIMEOUT |  10001 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| edn_sw_mode                                                 |    PASS |   1463 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_ast_rng_req_test                                |    PASS |    429 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_bypass_mode_health_test                         | TIMEOUT |  10001 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_csrng_test                                      | TIMEOUT |  10001 ms |        |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_edn_reqs_test                                   |   ERROR |    413 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_fw_observe_many_contiguous_test                 |    FAIL |   4097 ms |      6 |                              CHECK-STATUS-fail: DeadlineExceeded:["RVC";25] |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_fw_override_test                                | TIMEOUT |  10002 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_fw_ovr_test                                     |    PASS |    395 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_kat_test                                        |    FAIL |    284 ms |      6 |                            CHECK-fail: fail_count < kKatTestTimeoutAttempts |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| entropy_src_smoketest                                       | TIMEOUT |  10001 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| hmac_enc_idle_test                                          |    PASS |    393 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| hmac_enc_test                                               |    PASS |    371 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| hmac_secure_wipe_test                                       |    PASS |    405 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| hmac_smoketest                                              |    PASS |    580 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| otbn_irq_test                                               |    PASS |    282 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| otbn_isa_test                                               |    PASS |   1121 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| otbn_mem_scramble                                           |    FAIL |    539 ms |      6 |              CHECK-fail: Expecting at least 48 IMEM integrity errors; got 0 |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| otbn_randomness_test                                        |    PASS |    589 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| otbn_rsa_test                                               |    PASS |   2042 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
| otbn_smoketest_silicon_creator                              |    PASS |    286 ms |      6 |                                                                             |
+-------------------------------------------------------------+---------+-----------+--------+-----------------------------------------------------------------------------+
```

AlexJones0 commented on docs/opentitan/earlgrey.md:
```diff
@@ -155,8 +158,12 @@ See [`tools.md`](tools.md)
+  logged. `stderr` can be used to log the messages to the standard error stream instead of a file.
+
+* `-global ot-otbn.logasm=<true|false>` dumps executed instructions on OTBN core into the _logfile_
+  filename. Beware that this further slows down execution speed, which could likely result into
```
Nit: grammar (pre-existing before this PR).
```suggestion
  filename. Beware that this further slows down execution speed, which could likely result in the
```

rivos-eblot replied:
fixed

AlexJones0 reviewed (commented):
I'm not very familiar with the EDN and CSRNG and so can't provide a thorough review on the implementation specifics (I did find the CSRNG changes easier to review though); I've left some comments on things I noticed going over the code. On the whole the changes all look great, however, and it's nice to see more EDN/CSRNG tests passing. 

*Aside*: it is surprising to me to see `entropy_src_kat_test` failing, since I have it noted as passing - I went and ran it on `master` on OpenTitan and to my surprise it is failing! Going back to an older OpenTitan commit [`32b1337496`](https://github.com/lowRISC/opentitan/tree/32b1337496ebd8b76c9538a79c37d34122c35868) and running `./bazelisk.sh test //sw/device/tests:entropy_src_kat_test_sim_qemu_rom_with_fake_keys` it clearly is passing, so there has been a regression at some point, though not in this PR (so it shouldn't block merging). I would guess that the implementation of this test changed during the Multitop top-level-test porting efforts, in a way that is incompatible with the QEMU model (or maybe QEMU was bumped since then and a previous change broke this)?

rivos-eblot commented:
https://github.com/lowRISC/qemu/pull/153#discussion_r2018732351 not yet addressed (TBD next week).
Thanks for the review @AlexJones0 The EDN has been a large rewrite and comes as a very large change, which is not easy to read. There were some many changes to do that did not found a way to split it into coherent, small change commits.

rivos-eblot committed:
[ot] hw/opentitan: ot_random_src: rework interface and implementation

Get rid of the generation identifier. This mechanism had initially be implemented to verify
early EarlGrey entropy source constraint, quoting:

  "CSRNG may only be enabled if ENTROPY_SRC is enabled.
   Once disabled, CSRNG may only be re-enabled after ENTROPY_SRC has been disabled and
   re-enabled."

This constraint has been removed in recent EarlGrey implementations while Darjeeling
implementation did not follow this constraint, as it did not make use of the entropy source
IP.

Generation number is therefore removed; get_random_generation API is no longer needed, neither
is the genid argument of get_random_values. Error management can be simplified accordingly.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot commented:
> [#153 (comment)](https://github.com/lowRISC/qemu/pull/153#discussion_r2018732351) not yet addressed (TBD next week). 

It turns out that the generation identifier is no longer needed with the latest Entropy Src IP.
I'll remove the whole thing, as it simplifies both the API and the error management.

AlexJones0 reviewed (approved):
This looks good to me, thank you! I'm not familiar with the hardware enough to thoroughly review all the EDN changes but the code all seems reasonable and more tests are passing, so this is definitely an improvement.
